toolchain/check/handle_impl.cpp

```diff
@@ -428,6 +428,35 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
     FinishGenericDecl(context, impl_decl_id, impl_info.generic_id);
     impl_decl.impl_id = context.impls().Add(impl_info);
     lookup_bucket_ref.push_back(impl_decl.impl_id);
+
+    // Looking to see if there are any generic parameters on the `impl`
+    // declaration itself, not just generics inherited from it being written
+    // inside a generic context. If so, and the `impl` is concrete, the generic
+    // params will prevent the `impl` declaration from ever being used.
+    if (context.constant_values().Get(impl_info.self_id).is_concrete() &&
+        context.constant_values().Get(impl_info.constraint_id).is_concrete()) {
+      bool has_generic_param = false;
+      if (name.implicit_param_patterns_id.has_value()) {
+        for (auto inst_id :
+             context.inst_blocks().Get(name.implicit_param_patterns_id)) {
+          if (inst_id != SemIR::ErrorInst::SingletonInstId) {
+            has_generic_param = true;
+            break;
+          }
+        }
+      }
+      if (has_generic_param) {
+        // TODO: We should be able to diagnose an `impl` with a symbolic self
+        // and/or constraint, where it refers to some but not all of its
+        // generic parameters.
```

Comment:
Can we implement this by attempting to deduce the impl's parameters from its own self type and interface type? I think that should work if and only if all the parameters are deducible.

Reply:
Yeah! That works, at the cost of adding Converted instructions all over the place :upside_down_face: But I think we want to get rid of those from deduce anyway - or reuse the conversions at calls / impl lookups? For impl lookup maybe the only good option is to get rid of them.

Reply:
Yeah, those `converted` instructions are unfortunate. Maybe we can wrap this in `inst_block_stack().Push()` and `inst_block_stack().PopAndDiscard()` for now, just to keep the SemIR tidier?

Reply:
Oh that's an interesting idea, thanks, will do.